### PR TITLE
updated incentive

### PR DESCRIPTION
**Pull request by bendangnuksung**
Correction in incentive

### CODE REVIEW

**Comment by bendangnuksung** [comment]
changes made:
1.(contract/ModelRepository.sol) incentive calculation is not accurate because it is calculating the incentive without considering 'Model.targetError'. Its assumes the targetError =0.  Added incentiveCalculations() which considers the 'targetValue'  to derive  correct incentive.
2.(test/modelrepository.js) edited: targetValue, where targetValue < initialValue
3. (test/TestModelRepository) added: unit test for incentive calculations

File: contracts/ModelRepository.sol [hunk 1]
```diff
@@ -73,14 +73,22 @@ contract ModelRepository {
     models.push(newModel);
   }
 
+  function incentiveCalculate(uint bounty, uint totalError, uint solvedError) returns(uint total) {
```
Comment (axelhodler) [inline contracts/ModelRepository.sol]: can we mark this function `constant` see `getNumModels` then e.g. wallet software could call the function without a transaction if its just interested in learning "what would my incentive be?"

File: contracts/ModelRepository.sol [hunk 2]
```diff
@@ -73,14 +73,22 @@ contract ModelRepository {
...
+    total = ((bounty/totalError) * solvedError);
+    return total;
```
Comment (axelhodler) [inline contracts/ModelRepository.sol]: could do these two lines in one and remove the outer braces

File: contracts/ModelRepository.sol [hunk 3]
```diff
@@ -73,14 +73,22 @@ contract ModelRepository {
...
     grads[gradientId].newModelError = newModelError;
 
     Model model = models[grads[gradientId].modelId];
-    if (newModelError < model.bestError) {
-      uint incentive = ((model.bestError - newModelError) * model.bounty) / model.bestError;
+    if (newModelError < model.bestError && model.targetError < model.bestError) {
+      uint totalError = model.initialError - model.targetError;
+      uint solvedError = model.bestError - newModelError;
+
+      uint incentive = incentiveCalculate(model.bounty, totalError, solvedError);
```
Comment (axelhodler) [inline contracts/ModelRepository.sol]: `calculateIncentive` might seem like a better fit if we compare it with the other function names

File: test/TestModelRepository.sol [hunk 4]
```diff
@@ -13,5 +13,15 @@ contract TestModelRepository {
...
+  
+  function testIncentiveValue() {
+    ModelRepository repo = new ModelRepository();
+
+    //total error = initial error - target_error
+    //solved error = best_error - _new_model_error
+    //incentiveCalculate(bounty, total_error, solved_error)
 
+    Assert.equal(repo.incentiveCalculate(10000, 2000, 37), 185, "test case 1 ");
+    Assert.equal(repo.incentiveCalculate(123456789, 9, 4), 54869684, "test case 2");
```
Comment (axelhodler) [inline test/TestModelRepository.sol]: these two comments `test case 1/2` are unhelpful. maybe we can transfer the knowledge of line 20-22 into here
i would also do

```
bounty = 10000;
```

and the same for the other "magic numbers" to make it easier to read

File: test/modelrepository.js [hunk 5]
```diff
@@ -15,7 +15,7 @@ const firstModel = {
   id: 0,
   bountyInWei: 10000,
   initialError: 42,
-  targetError: 1337,
+  targetError: 10,
```
Comment (axelhodler) [inline test/modelrepository.js:18]: 😁 👍

File: test/modelrepository.js [hunk 6]
```diff
@@ -142,7 +142,7 @@ contract('ModelRepository - Evaluating Gradients', accounts => {
     const balanceAfterEvaluation = await balanceOf(patTheGradientProvider)
 
     assert.ok(balanceAfterEvaluation.valueOf() > balanceBeforeEvaluation.valueOf(),
-      'gradient provider was paid')
+      'gradient provider was not paid')
```
Comment (axelhodler) [inline test/modelrepository.js:158]: why not paid? `balanceAfterEvaluation.valueOf() > balanceBeforeEvaluation.valueOf()` seems to state the provider was paid.

Comment (bendangnuksung) [reply test/modelrepository.js:158]: Yes the provider is paid but shouldn't the 'message'  should be specifying the error message from the assertion error ? In this case, if  'GradientProvider' balance not increase then errorMessage be "GradientProvider not paid' ? 
I mean 'Message' is shown only when the 'assert()'  fails right?

File: test/modelrepository.js (hunk 5 above)
Comment (bendangnuksung) [reply test/modelrepository.js:18]: :)

File: test/TestModelRepository.sol (hunk 4 above)
Comment (bendangnuksung) [reply test/TestModelRepository.sol]: Okay. On it.

File: contracts/ModelRepository.sol (hunk 3 above)
Comment (bendangnuksung) [reply contracts/ModelRepository.sol]: Yes. Correct.

File: contracts/ModelRepository.sol (hunk 2 above)
Comment (bendangnuksung) [reply contracts/ModelRepository.sol]: Yes. Updating :+1:

File: contracts/ModelRepository.sol (hunk 1 above)
Comment (bendangnuksung) [reply contracts/ModelRepository.sol]: Yes, I will mark this function constant. Coming to your example, the software can call the function without transaction and calculate the incentive but it should do simple calculation to those variables (bestError,TargetError, InitialError and bounty) which are available and pass it to the function parameters. 
+ incentiveCalculate(bounty, initialError - targeterror, bestError - newError)

File: test/modelrepository.js (hunk 6 above)
Comment (bendangnuksung) [reply test/modelrepository.js]: I mean Message is shown only when the assert()  fails right?

Comment (axelhodler) [reply test/modelrepository.js:158]: true, my bad

**Comment by bendangnuksung** [comment]
Changes made on files mentioned above. Please check it. Thank you.

**Pull request merged**